futures-util/src/try_future/try_join.rs

```diff
@@ -0,0 +1,131 @@
+#![allow(non_snake_case)]
+
+use crate::future::{MaybeDone, maybe_done};
+use crate::try_future::{TryFutureExt, IntoFuture};
+use core::fmt;
+use core::mem::PinMut;
+use futures_core::future::{Future, TryFuture};
+use futures_core::task::{self, Poll};
+
+macro_rules! generate {
+    ($(
+        $(#[$doc:meta])*
+        ($Join:ident, <Fut1, $($Fut:ident),*>),
+    )*) => ($(
+        $(#[$doc])*
+        #[must_use = "futures do nothing unless polled"]
+        pub struct $Join<Fut1: TryFuture, $($Fut: TryFuture),*> {
+            Fut1: MaybeDone<IntoFuture<Fut1>>,
+            $($Fut: MaybeDone<IntoFuture<$Fut>>,)*
+        }
+
+        impl<Fut1, $($Fut),*> fmt::Debug for $Join<Fut1, $($Fut),*>
+        where
+            Fut1: TryFuture + fmt::Debug,
+            Fut1::Ok: fmt::Debug,
+            Fut1::Error: fmt::Debug,
+            $(
+                $Fut: TryFuture + fmt::Debug,
+                $Fut::Ok: fmt::Debug,
+                $Fut::Error: fmt::Debug,
+            )*
+        {
+            fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+                fmt.debug_struct(stringify!($Join))
+                    .field("Fut1", &self.Fut1)
+                    $(.field(stringify!($Fut), &self.$Fut))*
+                    .finish()
+            }
+        }
+
+        impl<Fut1, $($Fut),*> $Join<Fut1, $($Fut),*>
+        where
+            Fut1: TryFuture,
+            $(
+                $Fut: TryFuture<Error=Fut1::Error>
+            ),*
+        {
+            pub(super) fn new(Fut1: Fut1, $($Fut: $Fut),*) -> $Join<Fut1, $($Fut),*> {
+                $Join {
+                    Fut1: maybe_done(TryFutureExt::into_future(Fut1)),
+                    $($Fut: maybe_done(TryFutureExt::into_future($Fut))),*
+                }
+            }
+        }
+
+        impl<Fut1: TryFuture, $($Fut: TryFuture),*> $Join<Fut1, $($Fut),*> {
+            unsafe_pinned!(Fut1: MaybeDone<IntoFuture<Fut1>>);
+            $(
+                unsafe_pinned!($Fut: MaybeDone<IntoFuture<$Fut>>);
+            )*
```

Comment:
These `unsafe_pinned!` macro calls could be in the impl above. That's how we usually do it.

Reply:
Done.